### PR TITLE
Refactor encoding

### PR DESCRIPTION
**Pull request by wisecornelius**
## Problem

TL;DR: The lightning checkpointing requires some changes in the encoding. Further, the encoding is inefficient.

a) The pytorch-lightning checkpointing does not work properly, when passing binary objects into the init of `LightningModule`.
b) The dependency pattern between the Embedder, TargetCreator, ClassicModule and SequenceDataModule was not elegant.
c) DataLoader and TargetCreator performed tasks outside of their mission statement.
d) Data were encoded twice: inside the Embedder (the NN inputs) and inside the TargetCreator (targets). This should only be done once, and then features shifted.
e) There exist several of dependencies on information outside the scope of each class. E.g. the ClassicModule needed to know whether `OrdinalEncoderWithUnkown` is used to adapt the size of the heads.
f) The `CombinedEmbedder` as part of the NN, should not do data-preprocessing tasks.
g) The data outputted by the `SequenceLoader` were mixed between `np.ndarray` and `torch.Tensor`. Now everything that is either an input or a target of the NN will be delivered as torch.Tensor.
h) Some functions and classes were not placed properly in the last refactor and are out-of-place.

## Proposed changes

1) Introduce `Tokenizer` to do the tokenisation and trimming of sequences.
2) Introduce `FeatureDictionaryEncoder` to as encode the data.
3) The `TargetCreator` now only does what its name suggests: it shifts sequences.
4) The `SequenceLoader` now takes tokenizer, transform (encoder) and target_transform (target creator) as input.
5) The `CombinedEmbedder` is initialised inside the `ClassicModule` and now only embeds - no encoding.
6) Encoder, Embedder and ClassicModule now have methods to "officially" communicate with each other without accessing some attributes while making some adhoc changes etc. This should reduce the implicit assumptions that each component makes about the other drastically.
7) The DataModule will now be checkpointed properly.
8) Legacy code was removed and data utilities are better organised now: any util that can be used on its own (especially by the user) is now in the newly created `nl.utils.data`. For instance, the Tokenizer and Encoder.
9) Data are now processed as numeric numpy arrays instead of `object` which should speed up some operations siginficantly.
10) Speeds up training by approx 5%.

## Types of changes

What types of changes does your code introduce to neural-lifetimes?
_Put an `x` in the boxes that apply_

- [x] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Documentation Update (if none of the other choices apply)

## Checklist

- [x] I have read the [CONTRIBUTING](../CONTRIBUTING.md) doc
- [x] Description above provides context of the change
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] Unit tests for changes (not needed for documentation changes)
- [x] CI checks pass with my changes
- [ ] Bumping version in `setup.py` is an individual PR and not mixed with feature or bugfix PRs
- [x] Commits follow "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)"
- [x] Relevant documentation is updated including usage instructions

### CODE REVIEW

**Review by wisecornelius — COMMENTED** [review]
all temporary changes reversed.

**Review by pereplanellTW — APPROVED** [review]
